### DISCUSSION TITLE
How should I declare and keep the declared compatibility?

### DISCUSSION THREAD

**Discussion by daniel-a126** [microsoft/vscode-discussions#2437] [Extension Development QnA] [answered]
In multiple places (e.g. [here](https://code.visualstudio.com/api/get-started/extension-anatomy), [here](https://code.visualstudio.com/api/references/extension-manifest)) you can read about how to specify supported VSCode version (`engines/vscode`). This is typically paired with an example of `package.json` file showing also the dependencies (`devDependencies/@types/vscode`).

Something like this: 
```
  ...
  "engines": {
    "vscode": "^1.51.0"
  },
  ...
  "devDependencies": {
    "@types/vscode": "^1.51.0",
    ...
  }
```

While **engines** are clear - We typically want to support all VSCode version going forward (until a breaking major-version change comes) and therefore _carret_ seems to be appropriate, I can't really fully understand how the `@types/vscode` version specification is meant. If _carret_ is there, then with every update I will get newer and newer versions of it and I might easily end up using something that will break the declared backward compatibility.

Putting aside the fact that pretty much everybody updates VSCode to the latest version, it seems, that — technically speaking — I should keep the `@types/vscode` **hard-set to the oldest supported engine** (i.e. if `engines/vscode = ^1.51` then types should be `@types/vscode = 1.51` (no carret), am I right?

**Comment by robconery** [comment]
@isidorn can you help with this?

**Comment by alefragnani** [comment, marked as answer]
Hi @daniel-a126 ,

In essence, you are right. The safest approach is to hard-set `@types/vscode` to the oldest supported engine, and use `caret` in the `engines/vscode`, to be sure your extension will be _installable_ on future versions os VS Code.

But, in fact, this is not exclusive to VS Code, as any JS app could take care of the supported versions of their dependencies, and the `caret` symbol would avoid conflicts with breaking changes (major releases). 

BTW, I would say that you shouldn’t be too worried about _breaking changes_ in the VS Code API. The team works really hard to maintain a stable API, using deprecation notices whenever necessary. Being an extension developer myself since day 0 (almost 10 years), I remember just one episode.

Hope this helps
